File: frontend/erp-portfolio-web/src/types/pedidoCompra.ts
/**
 * =====================================================================
 * Arquivo....: pedidoCompra.ts
 * Versão.....: 1.0.0
 * Data.......: 23/09/2026
 * Descrição..: Tipos do módulo de Pedidos de Compra, espelhando os DTOs da
 *              API (PedidoCompraRespostaDto, PedidoCompraResumoDto,
 *              PedidoCompraCriacaoDto, PedidoCompraItemEntradaDto e
 *              PedidoCompraFiltroDto). Status chega e sai como TEXTO;
 *              reaproveita os mesmos valores de StatusPedido do Pedido de
 *              Venda (sem forma de pagamento: não há Contas a Pagar ainda).
 * ---------------------------------------------------------------------
 * Histórico de alterações:
 *   1.0.0 - 23/09/2026 - Criação do arquivo.
 * =====================================================================
 */

import type { StatusPedido } from './pedido'

export type { StatusPedido }

export interface PedidoCompraItem {
  id: number
  produtoId: number
  produtoNome: string
  sku: string
  unidade: string
  quantidade: number
  /** Custo do produto no momento em que o item foi adicionado (congelado). */
  precoUnitario: number
  descontoPercentual: number
  subtotal: number
}

export interface PedidoCompra {
  /** Também é o número do pedido de compra. */
  id: number
  fornecedorId: number
  fornecedorNome: string
  fornecedorDocumento: string
  /** Data/hora ISO 8601 em UTC. */
  dataPedido: string
  status: StatusPedido
  descontoPercentual: number
  subtotalItens: number
  valorTotal: number
  itens: PedidoCompraItem[]
}

/** Uma linha da listagem (sem os itens). */
export interface PedidoCompraResumo {
  id: number
  fornecedorId: number
  fornecedorNome: string
  dataPedido: string
  status: StatusPedido
  valorTotal: number
  quantidadeItens: number
}

/** Item enviado à API: sem preço (o servidor copia do Custo do produto). */
export interface PedidoCompraItemEntrada {
  produtoId: number
  quantidade: number
  descontoPercentual: number
}

/** Corpo do POST e do PUT (o PUT substitui o rascunho inteiro). */
export interface PedidoCompraEntrada {
  fornecedorId: number
  descontoPercentual: number
  itens: PedidoCompraItemEntrada[]
}

export interface PedidoCompraFiltro {
  /** Número do pedido (ex.: 12 ou #12) ou trecho do nome do fornecedor. */
  busca?: string
  status?: StatusPedido
  pagina: number
  tamanhoPagina: number
}

File: backend/ErpPortfolio.Api/Models/PedidoCompra.cs
// =====================================================================================
// Arquivo....: PedidoCompra.cs
// Versão.....: 1.0.0
// Data.......: 22/09/2026
// Descrição..: Entidade de domínio que representa um pedido de compra: cabeçalho com
//              fornecedor, status, desconto do pedido e o total calculado. Sem forma de
//              pagamento (SPEC.md, "Fora do escopo" — não há Contas a Pagar ainda).
// -------------------------------------------------------------------------------------
// Banco......: PostgreSQL - erp_portfolio_db
// Tabelas....: public.pedidos_compra (FK fornecedor_id -> public.fornecedores)
// Fontes.....: Mapeada em ErpPortfolioDbContext.PedidosCompra (EF Core / Npgsql).
//              Colunas: id, fornecedor_id, data_pedido, status, desconto_percentual,
//              valor_total (o mapeamento de nomes fica no DbContext).
//              O total é recalculado pelo serviço a cada gravação (Services/CalculoPedido.cs,
//              reaproveitado do Pedido de Venda).
// -------------------------------------------------------------------------------------
// Histórico de alterações:
//   1.0.0 - 22/09/2026 - Criação do arquivo.
// =====================================================================================

namespace ErpPortfolio.Api.Models;

public class PedidoCompra
{
    /// <summary>Também é o número do pedido de compra.</summary>
    public int Id { get; set; }

    public int FornecedorId { get; set; }

    public Fornecedor? Fornecedor { get; set; }

    /// <summary>Data/hora de criação em UTC; não é editável.</summary>
    public DateTime DataPedido { get; set; }

    public StatusPedido Status { get; set; } = StatusPedido.Rascunho;

    /// <summary>Desconto em % sobre a soma dos itens (0 a 100).</summary>
    public decimal DescontoPercentual { get; set; }

    /// <summary>Total do pedido em reais, gravado a cada salvamento.</summary>
    public decimal ValorTotal { get; set; }

    public List<PedidoCompraItem> Itens { get; set; } = [];
}
